djangochannelsrestframework/observer/model_observer.py

```diff
@@ -131,39 +140,46 @@ def post_change_receiver(self, instance: Model, action: Action, **kwargs):
         else:
             new_group_names = set(self.group_names_for_signal(instance=instance))
 
-        self.get_observer_state(instance).current_groups = new_group_names
-
-        # if post delete, new_group_names should be []
-
-        # Django DDP had used the ordering of DELETE, UPDATE then CREATE for good reasons.
-        self.send_messages(
-            instance, old_group_names - new_group_names, Action.DELETE, **kwargs
-        )
-        # the object has been updated so that its groups are not the same.
-        self.send_messages(
-            instance, old_group_names & new_group_names, Action.UPDATE, **kwargs
-        )
+        transaction.on_commit(partial(self._update_current_groups, instance, new_group_names))
 
-        #
-        self.send_messages(
-            instance, new_group_names - old_group_names, Action.CREATE, **kwargs
-        )
+        yield from self.generate_messages(instance, old_group_names, new_group_names, action, **kwargs)
 
-    def send_messages(
-        self, instance: Model, group_names: Set[str], action: Action, **kwargs
-    ):
-        if not group_names:
+    def _update_current_groups(self, instance, new_group_names):
+        self.get_observer_state(instance).current_groups = new_group_names
+    
+    def generate_messages(self, instance: Model, old_group_names: Set[str], new_group_names: Set[str], action: Action, **kwargs):
+        """
+        Generates messages for the given group names and action.
+        """
+        delete_group_names = old_group_names - new_group_names
+        if delete_group_names:
+            message_body = self.serialize(instance, Action.DELETE, **kwargs)
+            for group_name in delete_group_names:
+                yield {**message_body, "group": group_name}
+
+        update_group_names = old_group_names & new_group_names
+        if update_group_names:
+            message_body = self.serialize(instance, Action.UPDATE, **kwargs)
+            for group_name in update_group_names:
+                yield {**message_body, "group": group_name}
+
+        create_group_names = new_group_names - old_group_names
+        if create_group_names:
+            message_body = self.serialize(instance, Action.CREATE, **kwargs)
+            for group_name in create_group_names:
+                yield {**message_body, "group": group_name}
+
+    def send_prepared_messages(self, instance_pk):
+        """
+        Sends messages mapped to a specific instance after commit.
+        """
+        messages = self._instance_messages_mapping.pop(instance_pk, [])
```

Comment:
Good catch, since you are popping the messages here the fact that `send_prepared_messages` gets called multiple times does not effect things.